### PR TITLE
Eliminate redundant code blocks in modules and stages

### LINKED ISSUE #965: [FEA]: Identify and Remove Duplicate Code Blocks in Modules
### Is this a new feature, an improvement, or a change to existing functionality?

New Feature

### How would you describe the priority of this feature request

Medium

### Please provide a clear description of problem this feature solves

To facilitate future updates, it is recommended to extract common functionality from both stages and modules, as maintaining code with duplication in these components can pose challenges.

### Describe your ideal solution

Duplicate code sections present in both the module's code blocks should be consolidated into a common location accessible by both modules and stages.

### Describe any alternatives you have considered

_No response_

### Additional context

List of modules that requires changes
Morpheus Core Modules
- [x] [write_to_file](https://github.com/nv-morpheus/Morpheus/blob/branch-23.07/morpheus/stages/output/write_to_file_stage.py)
- [x] [serialize](https://github.com/nv-morpheus/Morpheus/blob/branch-23.07/morpheus/stages/postprocess/serialize_stage.py)
- [x] [filter_detections](https://github.com/nv-morpheus/Morpheus/blob/branch-23.07/morpheus/stages/postprocess/filter_detections_stage.py)
- [ ] [mlflow_model_writer](https://github.com/nv-morpheus/Morpheus/blob/branch-23.07/morpheus/modules/mlflow_model_writer.py)

Morpheus DFP Exmaple Modules (
Although not entirely duplicated, certain sections of the module's code blocks overlap with the stages in the given example)

- [ ] [All](https://github.com/nv-morpheus/Morpheus/tree/branch-23.07/examples/digital_fingerprinting/production/morpheus/dfp/modules)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I have searched the [open feature requests](https://github.com/nv-morpheus/Morpheus/issues?q=is%3Aopen+is%3Aissue+label%3A%22feature+request%22%2Cimprovement%2Cenhancement) and have found no duplicates for this feature request

### PR DESCRIPTION
**Pull request by bsuryadevara**
Eliminated redundant code blocks in both modules and stages, introducing controllers to enhance maintainability, and subsequently updated tests to align with these changes.

- file_to_df
- filter_detections
- mlflow_model_writer
- serializer
- write_to_file

Fixed preserve columns property issue.
closes #965 #1074

### CODE REVIEW

**Review by cwharris — APPROVED** [review]
LGTM. Glad to see this consolidation!

**Comment by bsuryadevara** [comment]
Changed to in progress. Fixing broken tests

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
A few general items:

- Move the controllers out of the `utils` directory and to their own directory `morpheus/controllers`
- Can you verify that there are no other modules that do not use a controller?
- Would it make more sense for some of the stages to just use the module

File: examples/ransomware_detection/common/feature_extractor.py [hunk 1]
```diff
@@ -381,8 +381,7 @@ def _extract_protections(self, x: pd.DataFrame, vadinfo_df_size: int, vadsinfo_s
         """
         page_execute_writecopy_count = 0
 
-        # pylint: disable=consider-iterating-dictionary
-        for protection in fc.PROTECTIONS.keys():
+        for protection, _ in fc.PROTECTIONS.items():
```
Comment (mdemoret-nv) [inline examples/ransomware_detection/common/feature_extractor.py:384]: See the documentation directly for what they are suggesting: https://pylint.readthedocs.io/en/latest/user_guide/messages/convention/consider-iterating-dictionary.html

You can just use:
```suggestion
        for protection in fc.PROTECTIONS:
```

File: examples/digital_fingerprinting/production/morpheus/dfp/utils/model_cache.py [hunk 2]
```diff
@@ -331,11 +331,10 @@ def load_model_cache(self, client: MlflowClient, reg_model_name: str, timeout: f
             logger.error("Deadlock when trying to acquire model cache lock")
             raise RuntimeError("Deadlock when trying to acquire model cache lock") from e
 
-    # pylint: disable=dangerous-default-value
     def load_user_model_cache(self,
                               user_id: str,
                               timeout: float,
-                              fallback_user_ids: typing.List[str] = []) -> UserModelMap:
+                              fallback_user_ids: typing.List[str] = None) -> UserModelMap:
```
Comment (mdemoret-nv) [inline examples/digital_fingerprinting/production/morpheus/dfp/utils/model_cache.py:337]: I assume that `None` is checked later on in `UserModelMap`? Can you confirm?

**Review by mdemoret-nv — APPROVED** [review]
Couple more items. Should be good to merge once the comments are resolved.

File: examples/digital_fingerprinting/production/morpheus/dfp/utils/model_cache.py (hunk 2 above)
Comment (bsuryadevara) [reply examples/digital_fingerprinting/production/morpheus/dfp/utils/model_cache.py]: Check is not performed at `UserModelMap`. I've relocated it to `ModelManager.load_user_model`, which is the starting point for loading the model.

**Pull request merged**